### DISCUSSION TITLE
Try to encrypt and decrypt using CMS in Langage C

### DISCUSSION THREAD

**Discussion by GillesT78** [openssl/openssl#25201] [Q&A]
I tried to encrypt and decrypt file using openssl library (langage C). I want to use CMS, SMIME and self-signed certificate.
As I don't succeed, I try with openssl binary, but I have also error on decryption
I create following script (on Windows)

SET OPENSSL=openssl.exe

Rem Create private key
%OPENSSL% genpkey -algorithm RSA -out private_key.pem -aes256

Rem Create a certificate signing request (CSR) using private key created
%OPENSSL% req -new -key private_key.pem -out certificate.csr

Rem Generate auto-signed certificate
%OPENSSL% x509 -req -days 365 -in certificate.csr -signkey private_key.pem -out certificate.crt

Rem Coinvert to format PKCS#12 (.p12)
%OPENSSL% pkcs12 -export -out certificate.p12 -inkey private_key.pem -in certificate.crt

Rem Extract public key of certificat
%OPENSSL% x509 -in certificate.crt -pubkey -noout > public_key.pem

Rem Crypt a file with CMS using public key
rem %OPENSSL% cms -encrypt -in file.txt -out file_crypted.cms -outform PEM public_key.pem
%OPENSSL% cms -encrypt -in file.txt -out file_crypted.cms -binary -outform PEM certificate.crt

Rem Decrypt the file le file 
%OPENSSL% cms -decrypt -in file_crypted.cms -out file_déchiffré.txt -inkey private_key.pem -recip certificate.crt

I have an error on decryption : "Error reading SMIME Content Info"
Why ?

Regards
Gilles

**Comment by mattcaswell** [comment]
The default input/output format is SMIME. However, in your encrypt command you explicitly set the outform to be PEM. Therefore when you decrypt the file you need to explicitly set the inform to be PEM as well. Add `-inform PEM` onto your decrypt command line.
